### PR TITLE
Osquerybeat: Rewrite queries scheduling  with extensions for scheduling and logging

### PR DESCRIPTION
**Pull request by aleksmaus**
## What does this PR do?

* Rewritten queries scheduling, leveraging custom plugins for osquery logging and configuration.
  1. Osquerybeat now registers the plugins for configuration and logging
  2. Osqueryd provides better scheduling for the queries, spreading the execution over time and state tracking.
  3. The configuration plugin is polled by osqueryd every 60 secs for the configuration changes (schedules queries). The interval for scheduled queries refresh is acceptable and can be increased to poll even less often.
  4. The queries are always scheduled as "snapshot" queries by osquerybeat configuration plugin.
  5. All osqueryd logs including the snapshot results are captured by the logging plugin.
* Refactored and cleaned up the osqueryd launcher and osqueryd client implementation.
 
## Why is it important?

Addresses https://github.com/elastic/security-team/issues/1061

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas

## Related issues

- Closes https://github.com/elastic/security-team/issues/1061

## Screenshots

The ```osquery_schedule``` is now populated with the scheduled queries stats 

Screen Shot 2021-05-10 at 10 27 36 AM

The scheduled queries and ad-hoc queries are working as expected

Screen Shot 2021-05-10 at 10 20 30 AM

Screen Shot 2021-05-10 at 10 19 50 AM

### CODE REVIEW

**Review by lykkin — APPROVED** [review]
:shipit: 🎉

File: x-pack/osquerybeat/beater/action_handler.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+	"github.com/elastic/beats/v7/x-pack/osquerybeat/internal/config"
+)
+
+type actionHandler struct {
+	inputType string
+	bt        *osquerybeat
+}
+
+func (a *actionHandler) Name() string {
```
Comment (scunningham) [inline x-pack/osquerybeat/beater/action_handler.go:20]: requires pointer?  not important, just curious

File: x-pack/osquerybeat/beater/action_handler.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+	return a.inputType
+}
+
+type actionData struct {
+	Query string
+	ID    string
+}
+
+func actionDataFromRequest(req map[string]interface{}) (ad actionData, err error) {
+	if req == nil {
```
Comment (scunningham) [inline x-pack/osquerybeat/beater/action_handler.go:30]: len(req) == 0 ?

File: x-pack/osquerybeat/beater/install.go [hunk 3]
```diff
@@ -15,7 +16,22 @@ import (
...
-func installOsquery(ctx context.Context, dir string) error {
+func installOsquery(ctx context.Context) error {
+	exefp, err := os.Executable()
+	if err != nil {
+		return err
+	}
+	exedir := filepath.Dir(exefp)
+
+	// Install osqueryd if needed
+	err = installOsqueryWithDir(ctx, exedir)
```
Comment (scunningham) [inline x-pack/osquerybeat/beater/install.go:27]: return installOsqueryWithDir(ctx, exedir) ?

**Review by scunningham — APPROVED** [review]

File: x-pack/osquerybeat/beater/action_handler.go (hunk 1 above)
Comment (aleksmaus) [reply x-pack/osquerybeat/beater/action_handler.go:20]: it is passed into the the agent client as interface so will escape to the heap no matter what, and it's only allocated once

File: x-pack/osquerybeat/beater/action_handler.go (hunk 2 above)
Comment (aleksmaus) [reply x-pack/osquerybeat/beater/action_handler.go]: sure

File: x-pack/osquerybeat/beater/install.go (hunk 3 above)
Comment (aleksmaus) [reply x-pack/osquerybeat/beater/install.go]: updated

**Pull request merged**